hamt/hamt.go

```diff
@@ -328,12 +331,17 @@ func (ds *Shard) loadChild(ctx context.Context, i int) (child, error) {
 	return c, nil
 }
 
-func (ds *Shard) setChild(i int, c child) {
+func (ds *Shard) setChild(i int, c *Shard) {
 	ds.children[i] = c
 }
 
 // Link returns a merklelink to this shard node
 func (ds *Shard) Link() (*ipld.Link, error) {
+	nodeType := ds.nodeType()
```

Comment:
`Link` is only called inside the `Node` method (hiding an implicit recursion) so we can just move this logic there and remove this function. However, if you think that it will generate too much confusion at this point in the PR we can iterate on it later.

Reply:
@schomatis If we move this `Link()` inside the `Node()`,  the `Node()` function may be too long.

Reply:
Agreed, we can leave this for another PR.